narwhals/_pandas_like/namespace.py

```diff
@@ -335,10 +336,117 @@ def __init__(
         self._root_names = root_names
         self._output_names = output_names
 
+    def when(self, condition: PandasLikeExpr) -> PandasChainedWhen:
+        return PandasChainedWhen(
+            self._call,  # type: ignore[arg-type]
+            condition,
+            depth=self._depth + 1,
+            implementation=self._implementation,
+            backend_version=self._backend_version,
+        )
+
     def otherwise(self, value: Any) -> PandasLikeExpr:
         # type ignore because we are setting the `_call` attribute to a
         # callable object of type `PandasWhen`, base class has the attribute as
         # only a `Callable`
         self._call._otherwise_value = value  # type: ignore[attr-defined]
         self._function_name = "whenotherwise"
         return self
+
+
+class PandasChainedWhen:
+    def __init__(
+        self,
+        above_when: PandasWhen | PandasChainedWhen,
+        condition: PandasLikeExpr,
+        depth: int,
+        implementation: Implementation,
+        backend_version: tuple[int, ...],
+        then_value: Any = None,
+        otherise_value: Any = None,
+    ) -> None:
+        self._implementation = implementation
+        self._depth = depth
+        self._backend_version = backend_version
+        self._condition = condition
+        self._above_when = above_when
+        self._then_value = then_value
+        self._otherwise_value = otherise_value
+
+        # TODO @aivanoved: this is way slow as during computation time this takes
+        # quadratic time need to improve this to linear time
+        self._condition = self._condition & (~self._above_when._already_set)  # type: ignore[has-type]
+        self._already_set = self._above_when._already_set | self._condition  # type: ignore[has-type]
```

Comment:
not sure I follow - rather than aiming for a complete solution from the start with a todo, could we start with the simplest possible case? this would already be enough to unblock running the rest of the tpc-h queries

i.e. to support `nw.when(...).then(...)` and `nw.when(...).then(...).otherwise(...)`, without chaining multiple of these together

it would also be easier to review like this, and we can then leave the chaining part for a follow-up

Reply:
@MarcoGorelli

this (commit)[https://github.com/narwhals-dev/narwhals/pull/588/commits/2f03bd0dbb4979bbec1e24ef8c4fcc504752e778] already implements the simple `nw.when(...).then(...)` and `nw.when(...).then(...).otherwise(...)`

And as of now other than the todo, we have the chaining as well

I wouldn't mind to revert to the above mentioned commit if we want to keep thing simple

Reply:
I can update the comment to be more detailed for sure

the reason this introduces a quadratic complexity for this operation is because I assign the expression to all previous masks to the current `ChainedWhen` mask and then call this expression in every `ChainedWhen` going down.

example:

```py
import pandas as pd

import narwhals as nw

data = {
    "a": [1, 2, 3, 4, 5],
}
df = nw.from_native(pd.DataFrame(data))

where_then_1 = nw.when(nw.col("a") == 1).then(7)
where_then_2 = where_then_1.when(nw.when(nw.col("a") == 3)).then(9).otherwise(19)

df.with_columns(where_then_2.alias("a_chained_when"))
```

When computing the mask for `when_then_2` it holds the expression for the mask of itself and `when_then_1`, `when_then_1` holds its own mask

When computing we call to see what the result of `when_then_1` is so we compute its mask with one call, cost of `1` for masks
When computing the result for `when_then_2`, we compute its condition's mask and the collective mask of all previous calls, which in this case is just the mask of `when_then_1`, so a cost of `1+1=2` for masks

The total cost for two chained calls is `3 = 2*3/2`

One could follow a similar logic to get that for `n`-chained calls we would compute masks a total of `n*(n+1)/2` times

Reply:
cool, thanks for explaining! sure, if you don't mind reverting and then keeping the additional complexity to a follow-up, I think we can probably merge this quite quickly